lib/cn/cn_tree.c

```diff
@@ -1805,15 +1809,22 @@ cn_comp_cleanup(struct cn_compaction_work *w)
     const bool kcompact = (w->cw_action == CN_ACTION_COMPACT_K);
     const bool spill = (w->cw_action == CN_ACTION_SPILL);
     const bool split = (w->cw_action == CN_ACTION_SPLIT);
+    const bool join = (w->cw_action == CN_ACTION_JOIN);
 
-    if (split) {
-        struct cn_tree_node *tn = w->cw_node;
+    if (split || join) {
+        struct cn_tree *tree = w->cw_tree;
+
+        mutex_lock(&tree->ct_ss_lock);
+        if (join) {
+            w->cw_join->tn_ss_joining = false;
+            w->cw_node->tn_ss_joining = false;
+        } else {
+            w->cw_node->tn_ss_splitting = false;
+        }
 
-        mutex_lock(&tn->tn_ss_lock);
-        tn->tn_ss_splitting = false;
-        atomic_dec(&tn->tn_tree->ct_split_cnt);
-        cv_broadcast(&tn->tn_ss_cv);
-        mutex_unlock(&tn->tn_ss_lock);
+        atomic_dec(&tree->ct_split_cnt);
+        cv_broadcast(&tree->ct_ss_cv);
+        mutex_unlock(&tree->ct_ss_lock);
```

Comment:
Both nodes of the join survive the join compaction job, althought the left-side node will have been merged into the right side and then removed from the route map.  the left side node is eventually deleted by csched within the context of the monitor thread.